rdchiral/template_extractor.py

```diff
@@ -93,18 +93,70 @@ def set_isotope_to_equal_mapnum(mol):
             a.SetIsotope(int(a.GetProp('molAtomMapNumber')))
-            
+
+def _invert_smarts_chirality_for_match(match, unmapped_only = True):
+    """
+    helper function for `invert_chirality_around_unmapped_ring_closure`
+    input is a re.match object that matches the portion of an atom token
+    where chirality is defined 
+    """
+    #only want to change unmapped token:
+    tetrahedral_token = match.group(1)
+    if ':' in tetrahedral_token and unmapped_only:
+        return tetrahedral_token
+
+    if '@@' in tetrahedral_token:
+        return tetrahedral_token.replace('@@', '@') 
+    elif '@' in tetrahedral_token:
+        return tetrahedral_token.replace('@', '@@')
+    else:
+        print ('Not a tetrahedral token:', tetrahedral_token)
+        return tetrahedral_token
+
+def invert_chirality_around_unmapped_ring_closure(smarts):
+    """
+    Return a smarts string with opposite chiral tags for atoms that precede a
+    ring closure token
+    """
+    return re.sub('(\[C@+H?\])(?=.?[1-9])', _invert_smarts_chirality_for_match, smarts)     
+
 def get_frag_around_tetrahedral_center(mol, idx):
     '''Builds a MolFragment using neighbors of a tetrahedral atom,
     where the molecule has already been updated to include isotopes'''
     ids_to_include = [idx]
     for neighbor in mol.GetAtomWithIdx(idx).GetNeighbors():
         ids_to_include.append(neighbor.GetIdx())
-    symbols = ['[{}{}]'.format(a.GetIsotope(), a.GetSymbol()) if a.GetIsotope() != 0\
-               else '[#{}]'.format(a.GetAtomicNum()) for a in mol.GetAtoms()]
+    # Get an initial smiles string for the fragment of the molecule -- 
+    # This string is used to get the correct chirality for the final, more general
+    # returned fragment.
+    # Likely a better way to do this, but chiral tags seem to be cleaned 
+    # by rdkit when obtained from a mol vs. from each atom individually
+    init_frag = Chem.MolFragmentToSmiles(mol, ids_to_include, isomericSmiles=True,
+                                    allBondsExplicit=True, allHsExplicit=True)
+    # assuming that for this task, only the chirality of mapped atoms in the fragment
+    # matters -- need to double check this 
+    # Get a list of each mapped atom token in the fragment
+    mapped_atom_tokens = re.findall('\[[0-9]+.*?:[0-9]+\]', init_frag)
+    map_num_to_chi = {int(re.findall('(?<=\[)[0-9]+', token)[0]):('').join(re.findall('@+H?', token)) for token in mapped_atom_tokens}
+    symbols = []
+    for a in mol.GetAtoms():
+        at_tag = ''
+        iso_tag = ''
+        chi_tag = ''
+        if a.GetIsotope() != 0:
+            at_tag = a.GetSymbol()
+            iso_tag = a.GetIsotope()
+
+            if iso_tag in map_num_to_chi.keys():
+                chi_tag = map_num_to_chi[iso_tag]
+        else:
+            at_tag = '#{}'.format(a.GetAtomicNum())
+
+        symbol = '[{}{}{}]'.format(iso_tag, at_tag, chi_tag)
+        symbols.append(symbol)
```

Comment:
@itai-levin  I think `chiral.copy_chirality` is better for preserving the chiral tags. It handles the cases where neighbor atoms are reordered.